src/nnet3/nnet-simple-component.cc

```diff
@@ -119,15 +119,21 @@ void DropoutComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
 
   BaseFloat dropout = dropout_proportion_;
   KALDI_ASSERT(dropout >= 0.0 && dropout <= 1.0);
-
+  // randomize the dropout matrix by row,
+  // i.e. [[1,1,1,1],[0,0,0,0],[0,0,0,0],[1,1,1,1],[0,0,0,0]]
   // This const_cast is only safe assuming you don't attempt
   // to use multi-threaded code with the GPU.
   const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(out);
-
-  out->Add(-dropout); // now, a proportion "dropout" will be <0.0
-  out->ApplyHeaviside(); // apply the function (x>0?1:0).  Now, a proportion "dropout" will
-                         // be zero and (1 - dropout) will be 1.0.
-
+  // now, a proportion "dropout" will be <0.0
```

Comment:
This wouldn't be compatible with regular uses of dropout.  You should add an option to the component.
And it's wasteful to do get a whole random matrix and only use one column; you should use a submatrix consisting of one column and randomly set just that.
Try again and I'll comment; this may take a few tries.